<div style="background-color:#182A1E; border:1px solid #587A59; border-left:8px solid #9BBF87; padding:22px 25px; border-radius:15px; box-shadow:0 6px 18px rgba(0,0,0,0.25);">

<div style="color:#9BBF87; font-size:13px; letter-spacing:2px; margin-bottom:7px;">
🌱 PRACTICAL DATA EXPLORATION
</div>

<h1 style="color:#E8F0DF; font-family:Georgia,serif; font-size:29px; margin:0;">
🧪 Hands-On Lab: Exploring a Real Dataset
</h1>

<p style="color:#AFC0AE; font-size:14px; margin:9px 0 0 0;">
🐼 Pandas &nbsp;•&nbsp; 🌍 CO₂ Emissions &nbsp;•&nbsp; 📊 Data Analysis
</p>

</div>

<div style="background-color:#1B2A20; border:1px solid #4F6F52; border-left:5px solid #7FA66A; padding:16px 20px; border-radius:11px;">

<h3 style="color:#B8D49F; font-family:Georgia,serif; font-size:20px; margin:0;">
🌿 Step 1 — Load & Inspect the Dataset
</h3>

<p style="color:#A8BDA8; font-size:14px; margin:7px 0 0 0;">
Load the CSV file and explore its shape, columns, and data types.
</p>

</div>

In [13]:
import pandas as pd

df = pd.read_csv("co2_emissions_data.csv")

df.head().style.set_table_styles([
    {
        'selector': 'th',
        'props': [
            ('background-color', '#4F6F52'),
            ('color', '#F1F7EC'),
            ('padding', '9px'),
            ('border', '1px solid #7FA66A'),
            ('font-weight', 'bold')
        ]
    },
    {
        'selector': 'td',
        'props': [
            ('background-color', '#1B2A20'),
            ('color', '#DDE8D6'),
            ('padding', '9px'),
            ('border', '1px solid #4F6F52')
        ]
    }
])

,country_code,country_name,year,value
0,ABW,Aruba,1960,11092.675000
1,ABW,Aruba,1961,11576.719000
2,ABW,Aruba,1962,12713.489000
3,ABW,Aruba,1963,12178.107000
4,ABW,Aruba,1964,11840.743000


In [14]:
lab_df = pd.read_csv("co2_emissions_data.csv")

print("\033[94mShape:\033[0m", lab_df.shape)
print("\033[92mColumns:\033[0m", lab_df.columns.tolist())
print("\033[93mData types:\033[0m")
print(lab_df.dtypes)

Shape: (13953, 4)
Columns: ['country_code', 'country_name', 'year', 'value']
Data types:
country_code        str
country_name        str
year              int64
value           float64
dtype: object


<div style="background-color:#1B2A20; border:1px solid #4F6F52; border-left:5px solid #7FA66A; padding:16px 20px; border-radius:11px;">

<h3 style="color:#B8D49F; font-family:Georgia,serif; font-size:20px; margin:0;">
🔍 Step 2 — Check & Handle Missing Values
</h3>

<p style="color:#A8BDA8; font-size:14px; margin:7px 0 0 0;">
Check every column for missing data and decide whether cleaning is necessary.
</p>

</div>

In [15]:
print("\033[94mMissing values per column:\033[0m")
lab_df.isnull().sum()

Missing values per column:


country_code    0
country_name    0
year            0
value           0
dtype: int64

<div style="background-color:#203326; border-left:4px solid #9BBF87; padding:14px 18px; border-radius:10px; margin-top:10px;">

<p style="color:#D5E8D8; font-size:15px; line-height:1.7; margin:0;">
🌱 <b style="color:#B8D49F;">Justification</b><br>
The dataset contains <b>zero missing values</b> across all columns
(<code>country_code</code>, <code>country_name</code>, <code>year</code>, and <code>value</code>).
Therefore, no filling or dropping is required. The data is already complete
and ready for analysis.
</p>

</div>

<div style="background-color:#1B2A20; border:1px solid #4F6F52; border-left:5px solid #7FA66A; padding:16px 20px; border-radius:11px;">

<h3 style="color:#B8D49F; font-family:Georgia,serif; font-size:20px; margin:0;">
🍃 Step 3 — Filter & Explore a Meaningful Subset
</h3>

<p style="color:#A8BDA8; font-size:14px; margin:7px 0 0 0;">
Filter the data using a meaningful condition and examine the resulting records.
</p>

</div>

In [16]:
lab_threshold = lab_df["value"].mean()
high_emission = lab_df[lab_df["value"] > lab_threshold]

print("\033[92mThreshold (mean value):\033[0m", lab_threshold)
print("\033[94mRows above threshold:\033[0m", high_emission.shape[0], "out of", lab_df.shape[0])
high_emission.head()

Threshold (mean value): 825498.3433543197
Rows above threshold: 1727 out of 13953


,country_code,country_name,year,value
423,ARB,Arab World,1996,843629.050531
424,ARB,Arab World,1997,884777.026039
425,ARB,Arab World,1998,889715.587754
426,ARB,Arab World,1999,901624.461624
427,ARB,Arab World,2000,939191.728434


In [17]:
print("Total rows in lab_df:", lab_df.shape[0])
print("Rows above threshold:", high_emission.shape[0])

Total rows in lab_df: 13953
Rows above threshold: 1727


<div style="background-color:#203326; border-left:4px solid #9BBF87; padding:14px 18px; border-radius:10px; margin-top:10px;">

<p style="color:#D5E8D8; font-size:15px; line-height:1.7; margin:0;">
🌍 <b style="color:#B8D49F;">Findings</b><br>
Filtering for records above the mean value resulted in
<b>1,727 out of 13,953 rows</b> with higher-than-average CO₂ emissions.
The subset highlights countries and aggregated regions with larger emission
values, including entities such as <b>"Arab World"</b>.
</p>

</div>

<div style="background-color:#1B2A20; border:1px solid #4F6F52; border-left:5px solid #7FA66A; padding:16px 20px; border-radius:11px;">

<h3 style="color:#B8D49F; font-family:Georgia,serif; font-size:20px; margin:0;">
📊 Step 4 — Group & Aggregate
</h3>

<p style="color:#A8BDA8; font-size:14px; margin:7px 0 0 0;">
Use <code>groupby</code> to calculate an aggregate statistic for each category
and interpret the results.
</p>

</div>

In [18]:
country_avg = lab_df.groupby("country_name")["value"].mean().sort_values(ascending=False)
print("\033[92mTop 10 countries by average CO2 value:\033[0m")
country_avg.head(10)

Top 10 countries by average CO2 value:


country_name
World                        2.195928e+07
High income                  1.112143e+07
IDA & IBRD total             1.111222e+07
OECD members                 1.089645e+07
IBRD only                    1.066460e+07
Low & middle income          1.065039e+07
Post-demographic dividend    1.061569e+07
Middle income                1.049463e+07
Upper middle income          8.038619e+06
Late-demographic dividend    7.788907e+06
Name: value, dtype: float64

<div style="background-color:#203326; border-left:4px solid #9BBF87; padding:14px 18px; border-radius:10px; margin-top:10px;">

<p style="color:#D5E8D8; font-size:15px; line-height:1.7; margin:0;">
🌱 <b style="color:#B8D49F;">Interpretation</b><br>
The top 10 entities with the highest average CO₂ values are mainly
<b>aggregated regions</b> rather than individual countries. This shows why
country-level records should be separated from regional or global groups
when performing precise comparisons.
</p>

</div>